Add a blog on multi-vector support in KNN (#2614)
opensearch-project/project-website: _posts/2024-02-23-enhanced-multi-vector-support-in-opensearch-knn.md

## Why is this important?
With the rise of Large Language Models (LLM), many users are turning to vector databases for indexing, storing, and retrieving information, particularly for building Retrieval-augmented Generation (RAG) systems. These models convert text to embeddings and preserve the semantic information in the embeddings. However, models have limitations on number of tokens to consider for embedding generation which requires document to be chunked and stored as multiple embeddings in a single document. Without multi-vector support, every chunk of the document needs to be treated as a separate entity, requiring duplication of document metadata across multiple chunk unless the user separates the document and its metadata. This duplication of metadata can result in significant storage and memory usage. Furthermore, users must independently devise a merging mechanism to ensure that only a single document is retrieved from multiple chunk during search. Multi-vector support simplifies the handling of large documents, alleviating these challenges.

> ## Understanding previous limitations

Reviewer: Insert an introductory sentence between headings. Something like "Here are some considerations to help you understand k-NN limitations prior to OpenSearch 2.12".

> ### Fewer Results

Reviewer: ```suggestion
### Fewer results
```

> In previous versions of OpenSearch, when searching for nested KNN (K-Nearest Neighbors) fields with a specified number of nearest neighbors (k value), the search might return fewer than k documents. This occurred because the search operated at the nested field level or chunk level rather than the document level.  In the worst scenario, its possible that k neighbors could be the chunks belonging to the single document, resulting in just one document in search results than desired. While increasing the k value could retrieve more documents, it introduced unnecessary search overhead.

Reviewer: ```suggestion
In versions of OpenSearch prior to 2.12, when you search for nested k-NN fields by providing a specified number of nearest neighbors (k value), the search might return fewer than k documents. In those OpenSearch versions, the search operates at the nested field level or chunk level rather than the document level.  In the worst case scenario, the retrieved k neighbors can be the chunks belonging to a single document, resulting in just one document in search results. While increasing the k value can retrieve more documents, it introduces unnecessary search overhead.
```

> Consider an example where there are three documents labeled 1, 2, and 3, each containing two vectors in the nested KNN field. Document 1 comprises vectors A and B, Document 2 includes vectors C and F, and Document 3 contains vectors D and E. When searching for the nested KNN field, if the two nearest vectors are A and B, both belonging to Document 1, it returns only Document 1, even if the k value is set to 2.

Reviewer: ```suggestion
For example, consider three documents labeled 1, 2, and 3, each containing two vectors in the nested k-NN field. Document 1 contains vectors A and B, Document 2 contains vectors C and F, and Document 3 contains vectors D and E. Let's say the two nearest vectors are A and B, both belonging to Document 1. When searching for the nested k-NN field, the search returns only Document 1, even if the k value is set to 2, as shown in the following image.
```

<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-before-1.png" alt="multi-vector-behavior-before" class="img-centered"/>

> ### Low Recall

Reviewer: ```suggestion
### Low recall
```

> The issue of fewer results mentioned earlier can also lead to low recall. In OpenSearch, an index consists of multiple shards, each containing multiple segments. Searches operate at the segment level, with results aggregated at the shard level. The results from all shards are aggregated and returned to the user finally. Consequently, if a single segment contains all of the top k documents but returns fewer than k documents, the final results, although containing k documents, would not represent the true top k documents. This is because the search algorithm operates at the segment level, potentially missing relevant documents from other segments that should have been included in the top k results.

Reviewer: ```suggestion
Returning fewer results mentioned earlier can also lead to low recall. In OpenSearch, an index consists of multiple shards, each containing multiple segments. Search operates at the segment level, with results aggregated at the shard level. After this, the results from all shards are aggregated and returned to the user. Consequently, if a single segment contains all of the top k documents but returns fewer than k documents, the final results will contain k documents but will not represent the true top k documents. This is because the search algorithm operates at the segment level, potentially missing relevant documents from other segments that should have been included in the top k results.
```

> Consider an illustrative example to delve into the issue further. Suppose there are 7 indexed documents, each containing either one or two vectors. Let's assume that, when searching the nested vector field, the order of vectors from nearest to farthest is as follows: Vector A, B, C, D, E, F, G, H, I, and J. The search occurs at the segment level, and with a k value of 2, only document 1 is returned from segment 1. Consequently, the final aggregated result comprises document 1 and document 3, whereas the expected outcome should include document 1 and document 2.

Reviewer: ```suggestion
To illustrate this, consider the following example. Suppose there are 7 indexed documents, each containing either one or two vectors. Let's assume that, when searching the nested vector field, the order of vectors from nearest to farthest is as follows: A, B, C, D, E, F, G, H, I, and J. The search occurs at the segment level, and with a k value of 2, only document 1 is returned from segment 1. Consequently, the final aggregated result contains document 1 and document 3, whereas the expected result should include document 1 and document 2, as shown in the following image.
```

<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-before-2.png" alt="multi-vector-behavior-before-cluster-level" class="img-centered"/>

> ## Improvements in OpenSearch 2.12

Reviewer: ```suggestion
## Improvements in OpenSearch 2.12

```

> In the latest version, OpenSearch leverages document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector closest to the query vector and discards the farther one. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.

Reviewer: ```suggestion
In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector closest to the query vector and discards the farther one. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.
```

> Let’s take a look at an example with the diagrams below. Consider two documents, Document 1 and Document 2. Document 1 contains vectors A and B, while Document 2 contains vectors C and D (Fig. 1). During the search, vector A is found and added to the search queue (Fig. 2). Subsequently, vector B is encountered. Since both vectors A and B belong to Document 1, their distances are compared. The distance between the query vector and vector B is 0.9, which is less than the distance between the query vector and vector A, which is 1.0. As a result, vector A is removed from the search queue, and vector B is added (Fig. 3).

Reviewer: ```suggestion
Let’s take a look at an example presented in the following diagrams. Consider two documents, Document 1 and Document 2. Document 1 contains vectors A and B, while Document 2 contains vectors C and D (Fig. 1). During the search, vector A is found and added to the search queue (Fig. 2). Subsequently, vector B is encountered. Since both vectors A and B belong to Document 1, their distances are compared. The distance between the query vector and vector B is 0.9, which is less than the distance between the query vector and vector A (1.0). As a result, vector A is removed from the search queue, and vector B is added (Fig. 3).
```

> Moving forward, vector D is discovered. Since vector D belongs to Document 2, whose vector is not in the search queue, it is added (Fig. 4). Then, vector C is found. As vector C belongs to Document 2, whose vector is already in the search queue, the distances between the query vector and vectors C and D are compared. The distance between the query vector and vector D (1.1) is less than that between the query vector and vector C (1.2), so vector D remains in the search queue (Fig. 5).

Reviewer: ```suggestion
After that, vector D is discovered. Since vector D belongs to Document 2, whose vector is not in the search queue, vector D is added to the queue (Fig. 4). Then, vector C is found. Because vector C belongs to Document 2, whose vector is already in the search queue, the distances between the query vector and vectors C and D are compared. The distance between the query vector and vector D (1.1) is less than the distance between the query vector and vector C (1.2), so vector D remains in the search queue (Fig. 5).
```

In the end, both Document 1 and Document 2 are returned, and their scores are calculated based on the distances between the query vector and the vectors collected in the search queue.

<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-after-1.png" alt="multi-vector-behavior-after" class="img-centered"/>

> Now, even with a k value set to 2, the search will return 2 documents, regardless of whether the 2 nearest vectors belong to one or multiple documents.  Additionally, this enhancement improves recall for users, as each segment now returns the k nearest documents instead of just the k nearest vectors of documents.

Reviewer: ```suggestion
Now, even with a k value set to 2, the search returns two documents, regardless of whether the two nearest vectors belong to one or multiple documents.  Additionally, this enhancement improves recall, because each segment now returns the k nearest documents instead of just the k nearest vectors belonging to documents.
```

> ## How to use nested field to store multi-vector

Reviewer: ```suggestion
## How to use a nested field to store a multi-vector

```

> To utilize nested fields for storing multi-vectors, ensure that your OpenSearch cluster has the KNN plugins installed. Now, let's delve into the process of creating a KNN index with nested fields and conducting searches on it.

Reviewer: I don't think we should call out installing the k-NN plugin because it is installed by default and presumably the users reading this blog will be familiar with k-NN already.
```suggestion
Let's delve into the process of creating a k-NN index with nested fields and conducting searches on it.
```

> Begin by creating a KNN index, setting the knn value to true in the index settings. Additionally, define the knn_vector type within the nested type. All other parameters for the knn_vector type remain the same as those for a normal knn_vector type.

Reviewer: ```suggestion
Begin by creating a k-NN index by setting the `knn` value to `true` in the index settings. Additionally, set the `type` to `knn_vector` within the nested field. All other parameters for the `knn_vector` remain the same as those for a regular `knn_vector` type:

```

> ```

Reviewer: ````suggestion
```json
````

PUT my-knn-index
{
  "settings": {
    "index": {
      "knn": true
    }
  },
  "mappings": {
    "properties": {
      "my_vectors": {
        "type": "nested",
        "properties": {
          "my_vector": {
            "type": "knn_vector",
            "dimension": 2,
            "method": {
              "name": "hnsw",
              "space_type": "l2",
              "engine": "faiss"
            }
          }
        }
      }
    }
  }
}
```

> Next, insert your data. The number of data entries within a nested field is not fixed; you can index a different number of nested field items for each document. For instance, in this example, three nested field items were added for document 1, and two for document 2.

Reviewer: ```suggestion
Next, insert your data. The number of data entries within a nested field is not fixed; you can index a different number of nested field items for each document. For instance, in this example, document 1 contains three nested field items and document 2 contains two nested field items:

```

```
PUT _bulk?refresh=true
{ "index": { "_index": "my-knn-index", "_id": "1" } }
{"my_vectors":[{"my_vector":[1,1]},{"my_vector":[2,2]},{"my_vector":[3,3]}]}
{ "index": { "_index": "my-knn-index", "_id": "2" } }
{"my_vectors":[{"my_vector":[10,10]},{"my_vector":[20,20]}]}
```

> When ready to search the data, note that the query structure differs slightly from a normal KNN search. Your query should be wrapped by a nested query with the specified path. Moreover, your field name should encompass both the nested field name and the knn_vector field name, concatenated with a dot.

Reviewer: ```suggestion
When you search the data, note that the query structure differs slightly from a regular k-NN search. Wrap your query in a nested query with the specified path. Additionally, your field name should specify both the nested field name and the `knn_vector field name`, separated by a dot (in the following example, `my_vectors.my_vector`):

```

> ```

Reviewer: ````suggestion
```json
````

GET my-knn-index/_search
{
  "query": {
    "nested": {
      "path": "my_vectors",
      "query": {
        "knn": {
          "my_vectors.my_vector": {
            "vector": [1,1],
            "k": 2
          }
        }
      }
    }
  }
}
```

> With a k value of 2, you'll retrieve two documents instead of one, even if the three nearest vectors belong to document 1.

Reviewer: ```suggestion
When specifying a k value of 2, you'll retrieve two documents instead of one, even if the three nearest vectors belong to document 1:

```

Author: Is it `:`? not `.` at the end?

> ```

Reviewer: ````suggestion
```json
````

                                1
                            ]
                        },
                        {
                            "my_vector": [
                                2,
                                2
                            ]
                        },
                        {
                            "my_vector": [
                                3,
                                3
                            ]
                        }
                    ]
                }
            },
            {
                "_index": "my-knn-index",
                "_id": "2",
                "_score": 0.006134969,
                "_source": {
                    "my_vectors": [
                        {
                            "my_vector": [
                                10,
                                10
                            ]
                        },
                        {
                            "my_vector": [
                                20,
                                20
                            ]
                        }
                    ]
                }
            }
        ]
    }
}

> ```

Reviewer: ````suggestion
```

````

> ## Summary

Reviewer: ```suggestion
## Summary

```

> In OpenSearch 2.12, users can now obtain diverse search results from the KNN index, even when multiple nearest vectors belong to just a few documents. This represents a significant stride towards establishing OpenSearch as a competitive vector database. Additional enhancements to support multi-vector functionality, such as inner hit support[[#1447](https://github.com/opensearch-project/k-NN/issues/1447)] and automatic chunking[[#548](https://github.com/opensearch-project/neural-search/issues/548)], are in the pipeline. If you require these features, please give a +1 to the corresponding GitHub issue. Furthermore, feel free to submit new issues on GitHub for any other ideas or requests regarding OpenSearch KNN functionality.

Reviewer: Is "diverse" the right adjective here? Should it be "more accurate"?
```suggestion
In OpenSearch 2.12, you can now obtain diverse search results from a k-NN index, even when multiple nearest vectors belong to just a few documents. This represents a significant stride towards establishing OpenSearch as a competitive vector database. Additional enhancements to support multi-vector functionality, such as inner hit support[[#1447](https://github.com/opensearch-project/k-NN/issues/1447)] and automatic chunking[[#548](https://github.com/opensearch-project/neural-search/issues/548)], are in the pipeline. If you want to see these features implemented, please upvote the corresponding GitHub issue. As always, feel free to submit new issues for any other ideas or requests regarding the OpenSearch k-NN functionality in the [k-NN repository](https://github.com/opensearch-project/k-NN).
```

## Understanding previous limitations

There are two limitations in k-NN with nested field prior to OpenSearch 2.12. 

### Fewer results
In previous versions of OpenSearch, when searching for nested k-NN (K-Nearest Neighbors) fields with a specified number of nearest neighbors (k value), the search might return fewer than k documents. This occurred because the search operated at the nested field level or chunk level rather than the document level. In the worst scenario, its possible that k neighbors could be the chunks belonging to the single document, resulting in just one document in search results than desired. While increasing the k value could retrieve more documents, it introduced unnecessary search overhead.

> For example, consider three documents labeled 1, 2, and 3, each containing two vectors in the nested k-NN field. Document 1 contains vectors A and B, Document 2 contains vectors C and F, and Document 3 contains vectors D and E. Let's say the two nearest vectors are A and B, both belonging to Document 1. When searching for the nested k-NN field, the search returns only Document 1, even if the k value is set to 2, as shown in the following image.

Reviewer 2: ```suggestion
For example, consider three documents labeled 1, 2, and 3, each containing two vectors in the nested k-NN field. Document 1 contains vectors A and B, document 2 contains vectors C and F, and document 3 contains vectors D and E. Let's say the two nearest vectors are A and B, both belonging to document 1. When searching for the nested k-NN field, the search returns only document 1, even if the k value is set to 2, as shown in the following image.
```

To illustrate this, consider the following example. Suppose there are 7 indexed documents, each containing either one or two vectors. Let's assume that, when searching the nested vector field, the order of vectors from nearest to farthest is as follows: A, B, C, D, E, F, G, H, I, and J. The search occurs at the segment level, and with a k value of 2, only document 1 is returned from segment 1. Consequently, the final aggregated result contains document 1 and document 3, whereas the expected result should include document 1 and document 2, as shown in the following image.

<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-before-2.png" alt="multi-vector-behavior-before-cluster-level" class="img-centered"/>

## Improvements in OpenSearch 2.12

> In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector closest to the query vector and discards the farther one. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.

Author: There will be only two vector to choose from at a time. We choose nearer one and drop the farther one and this can happen multiple times for the same document.

> In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector closest to the query vector and discards the farther one. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.

Author: Updated as 
```
In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector nearer to the query vector and discards the farther one out of the two. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.
```

> In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector closest to the query vector and discards the farther one. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.

Author: Updated as
`In version 2.12, OpenSearch uses document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch keeps only the nearest vector to the query vector for each document. The distance value of the selected vector is then converted back to an OpenSearch search score, ultimately becoming the document's score.`
